File: Journey-to-silicon-valley/Week_1/day-2/maxProfit.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minimum = prices[0]
        index = 0
        profit = 0

        for i in range(1, len(prices)):
            if prices[i] < minimum:
                minimum = prices[i]
                index = i
            
            new_profit = max(prices[index:]) - minimum
            profit = profit if profit > (new_profit) else new_profit

        return profit



print(Solution().maxProfit( prices = [7,1,5,3,6,4]))
print(Solution().maxProfit( prices = [7,6,4,3,1]))
print(Solution().maxProfit( prices = [2,4,1]))

File: Journey-to-silicon-valley/Week_1/day-2/maxProfit.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        int maxProfit(vector<int>& prices) {
            int buyingPrice = prices[0];
            int profit = 0;

            for (size_t i = 0; i < prices.size(); i++)
            {
                if(buyingPrice > prices[i]){
                    buyingPrice = prices[i];
                }

                if(profit < prices[i] - buyingPrice ){
                    profit = prices[i] - buyingPrice;
                }
            }
            return profit;
            

        }
    };